### ISSUE TITLE
kernelinfo doesn't compile for linux_osi

### ISSUE THREAD

**Issue by bridgeythegeek** [panda-re/panda#194]
Trying to compile `kernelinfo` as per [the instructions](https://github.com/panda-re/panda/blob/master/panda/plugins/osi_linux/USAGE.md), but get errors:

```
ake -C /lib/modules/4.4.0-81-generic/build M=/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo modules
make[1]: Entering directory '/usr/src/linux-headers-4.4.0-81-generic'
  CC [M]  /home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.o
/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.c: In function ‘init_module’:
/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.c:56:132: error: ‘struct vfsmount’ has no member named ‘mnt_parent’
 RINT_OFFSET(structp, memb, cfgname) printk(KERN_INFO "%s.%s_offset = %d", cfgname, cp_memb(#memb), (int)((void *)&(structp->memb) - (
                                                                                                                           ^
/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.c:138:2: note: in expansion of macro ‘PRINT_OFFSET’
  PRINT_OFFSET(vfsmnt_p, mnt_parent,  "fs");
  ^
/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.c:56:132: error: ‘struct vfsmount’ has no member named ‘mnt_mountpoint’
 RINT_OFFSET(structp, memb, cfgname) printk(KERN_INFO "%s.%s_offset = %d", cfgname, cp_memb(#memb), (int)((void *)&(structp->memb) - (
                                                                                                                           ^
/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.c:139:2: note: in expansion of macro ‘PRINT_OFFSET’
  PRINT_OFFSET(vfsmnt_p, mnt_mountpoint, "fs");
  ^
scripts/Makefile.build:264: recipe for target '/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.o' failed
make[2]: *** [/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo/kernelinfo.o] Error 1
Makefile:1420: recipe for target '_module_/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo' failed
make[1]: *** [_module_/home/btg/build-panda/panda/panda/plugins/osi_linux/utils/kernelinfo] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.4.0-81-generic'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2
```

Have successfully installed kernel headers with:
```
sudo apt-get install linux-headers-$(uname -r)
```

**Comment by m000** [comment]
Kernel internals change more often than you'd expect. Last known working kernel is 3.16. If you want to add support for more a more recent kernel, it'd be great. I can give you feedback if you are willing to undertake this.
If that would work for you, I can also share a Debian 7 image which runs a compatible kernel. Just send me an email.

See also: #181

**Comment by bridgeythegeek** [comment]
Thanks! I'm happy to have a look at adding support for a more recent kernel if there's some guidance you can offer?

**Comment by m000** [comment]
Overview of the steps for adding support for a new kernel:
- You should start by reading [kernel_structs.html](https://github.com/panda-re/panda/blob/master/panda/plugins/osi_linux/kernel_structs.html). That is a sketch of the structures you have to traverse as of Linux 3.16.
- Then see the error you got when compiling the kernelinfo module, and find the missing member in the sketch. That is your starting point.
- Browse the source of the kernel you are targeting on github, and try to identify how the equivalent information is currently represented. Commit logs or kernel change logs may help you.
- Add to kernel_structs.html a sketch for the structures layout in the new kernel.
- Modify kernelinfo to extract the offsets you want. Use appropriate ifdefs so that compilation for older kernels will not break.
- Contact me again to discuss how it would be best to implement the runtime traversal of the kernel structures (modifications in the osi_linux plugin itself).

**Comment by antverdovsky** [comment]
Just wondering, has any progress has been made on this?

**Comment by moyix** [comment]
Not on our end. @bridgeythegeek, did you have any luck?

**Comment by bridgeythegeek** [comment]
Not yet, but it's my job for this weekend, so hopefully will report back then!

**Comment by bridgeythegeek** [comment]
Sadly, this looks like it's going to be beyond my current skill set. There's a lot of background stuff that I simply don't know and unfortunately I don't have the time to dig in right now. Sorry all.

**Comment by bridgeythegeek** [comment]
All right, it's amazing what distracts you on a long train journey. Comparing to kernel 4.12.10 (current stable), it looks like the only thing from the sketch that's changed is `struct vfsmount` - which fits perfectly with my error messages.

From the git commits, it looks like there were a run of changes in January 2012 that made the changes. Fortunately, it looks like the two members we're interested in (`mnt_parent` and `mnt_mountpoint`) were simply moved from `struct vfsmount` to `struct mount`; see [this commit](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/commit/?h=v4.10&id=3376f34fff5be9954fd9a9c4fd68f4a0a36d480e) for changes to `mnt_parent` as an example.

It might just be within my skill set after all. I'll see what I can achieve on the return train journey.

**Comment by bridgeythegeek** [comment]
Ok, I think I've done all right!

The following  shows the changes I made to `kernelinfo.c`:

#### 1) Add an include for `mount.h` from the `fs` folder of the kernel source.
`#include "../linux-stable/fs/mount.h"`

I'm hoping this isn't quite the nightmare it seems to be. The move from `vfsmount` to `mount` means you need to include `/fs/mount.h` (from the Kernel source). However, this file isn't provided when you do `sudo apt-get install linux-headers-$(uname -r)`. I couldn't find a package that would provide it so I had to `git clone` the whole Linux Kernel source, hence the `"../linux-stable/fs/mount.h"`.

Am I missing something?!

#### 2) Add a `struct` for `mount`.
`struct mount mountstruct;`
 
#### 3) Add a pointer.
`struct mount *mnt_p;`

#### 4) Give the pointer a value. 
`mnt_p = &mountstruct;`
 
#### 5) Remove the lines that no longer work.
`PRINT_OFFSET(vfsmnt_p,	mnt_parent,		"fs");`
`PRINT_OFFSET(vfsmnt_p,	mnt_mountpoint, "fs");`

#### 6) Replace them with lines that reference the new structure.
`PRINT_OFFSET(mnt_p,	mnt_parent,		"fs");`
`PRINT_OFFSET(mnt_p,	mnt_mountpoint, "fs");`

#### 7) In `../linux-stable/fs/mount.h`, remove the layout randomization because we don't need it and it adds another hurdle. From both `struct mnt_namespace` and `struct mount`.
`}; //__randomize_layout;`

Now when I `make`, it seems to go fine:
```
$ make
make -C /lib/modules/4.4.0-93-generic/build M=/home/btg/src/kernelinfo modules
make[1]: Entering directory '/usr/src/linux-headers-4.4.0-93-generic'
  CC [M]  /home/btg/src/kernelinfo/kernelinfo.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /home/btg/src/kernelinfo/kernelinfo.mod.o
  LD [M]  /home/btg/src/kernelinfo/kernelinfo.ko
make[1]: Leaving directory '/usr/src/linux-headers-4.4.0-93-generic'
```
And, with `insmod`, as expected:
```
$ sudo insmod kernelinfo.ko
insmod: ERROR: could not insert module kernelinfo.ko: Operation not permitted
```
So, next steps:

1. Can anybody answer my question about requiring the kernel source to compile?
1. Is it sensible to do some Linux header version checking in `kernelinfo.c' (e.g. `#if LINUX_VERSION_CODE <= KERNEL_VERSION(X,Y,Z)`) or is there a more sensible way to do this (before I submit a pull request).

**Comment by bridgeythegeek** [comment]
Oh. It turns out, you literally just need the `mount.h` file of `/fs/mount.h`. I thought there'd be crazy dependencies, but evidently not. So I just copied `fs/mount.h` into the `kernelinfo` folder and `make` was quite happy!

**Comment by bridgeythegeek** [comment]
Huh. If `struct mount` is using `__randomize_layout`, does that mean some kind of ASLR is going on with the struct. Meaning the kernel offsets will have to be regenerated on every boot?

**Comment by m000** [comment]
I believe it is compile-time. See https://lwn.net/Articles/723997/

As such, it adds another level of complexity but one should be able to
address it it with some additional engineering effort. For research
purposes it may not worth the trouble. But it could make a nice student
project.

Any clues if this randomization is enabled by default in any distro?

M.


2017-09-02 17:54 GMT+02:00 Bridgey the Geek <notifications@github.com>:

**Comment by StablelJay** [comment]
@bridgeythegeek  I fialed to compile linux kernelinfo on Ubuntu 16.04 VM with your method. Since you have made it on Debian 7 VM, could it work for Ubuntu VM, or just support Debian system?

**Comment by bridgeythegeek** [comment]
@StablelJay The particular distro (Debian vs Ubuntu) shouldn't make a difference, it's the kernel structure. What error message did you get? Also, what kernel do you have? (`uname -a`)

**Comment by m000** [comment]
I believe this can be closed after #354 .

**Comment by moyix** [comment]
Hopefully this is fixed now, closing.
